src/sysdolphin/baselib/class.c

```diff
@@ -142,3 +143,338 @@ HSD_MemoryEntry* GetMemoryEntry(s32 idx)
+{
+    return object->parent.class_info;
+}
+
+inline HSD_ClassInfo* HSD_PushClassInfo(HSD_ClassInfo* class_info)
+{
+    HSD_ClassInfo* ret;
+    return ret = class_info;
+}
+
+BOOL func_80382400(HSD_Obj* object, HSD_ClassInfo* class_info)
+{
+    HSD_ClassInfo* var_r29;
+    HSD_ClassInfo* var_r28;
+
+    assert_line(0x249, object);
+    assert_line(0x24A, class_info);
+    var_r29 = HSD_GetClassInfo(object);
+    !var_r29;
+    var_r28 = HSD_PushClassInfo(class_info);
+    if (!(var_r28->head.flags & 1)) {
+        var_r28->head.info_init();
+    }
+    if (var_r29->head.obj_size != var_r28->head.obj_size) return FALSE;
+    while (var_r29->head.parent != NULL && var_r29->head.parent->head.obj_size == var_r29->head.obj_size) {
+        var_r29 = var_r29->head.parent;
+    }
+    while (var_r28->head.parent != NULL && var_r28->head.parent->head.obj_size == var_r28->head.obj_size) {
+        var_r28 = var_r28->head.parent;
+    }
+    if (var_r29 == var_r28) {
+        var_r29->head.nb_exist--;
+        class_info->head.nb_exist++;
+        if (class_info->head.nb_exist > class_info->head.nb_peak)
+        {
+            class_info->head.nb_peak = class_info->head.nb_exist;
+        }
+        object->parent.class_info = class_info;
+        return TRUE;
+    }
+    return FALSE;
+}
+
+BOOL hsdIsDescendantOf(HSD_ClassInfo* info, HSD_ClassInfo* p)
+{
+    HSD_ClassInfo* var_r31;
+
+    if (info == NULL || p == NULL) {
+        return FALSE;
+    }
+    var_r31 = var_r31 = info;
+    if (!(info->head.flags & 1)) {
+        var_r31->head.info_init();
+    }
+    if (!(p->head.flags & 1)) {
+        p->head.info_init();
+    }
+    while (var_r31 != NULL) {
+        if (var_r31 == p) {
+            return TRUE;
+        }
+        var_r31 = var_r31->head.parent;
+    }
+    return FALSE;
+}
+
+BOOL hsdObjIsDescendantOf(HSD_Obj* o, HSD_ClassInfo* p)
+{
+    HSD_ClassInfo* info;
+
+    if (o == NULL || p == NULL) {
+        return FALSE;
+    }
+    info = o->parent.class_info;
+    if (!(p->head.flags & 1)) {
+        p->head.info_init();
+    }
+    while (info != NULL) {
+        if (info == p) {
+            return TRUE;
+        }
+        info = info->head.parent;
+    }
+    return FALSE;
+}
+
+// (name based on KAR address)
+void func_80420FBC(HSD_ClassInfo* class_info, s32 arg1, s32 arg2)
+{
+    class_info->head.flags = class_info->head.flags & ~arg2 | arg1;
+}
+
+void ForgetClassLibraryReal(HSD_ClassInfo* class_info)
+{
+    HSD_ClassInfo* cur = class_info->head.child;
+    HSD_ClassInfo* next;
+    while (cur != NULL) {
+        next = cur->head.next;
+        cur->head.next = NULL;
+        ForgetClassLibraryReal(cur);
+        cur = next;
+    }
+    class_info->amnesia(class_info);
+    class_info->head.child = NULL;
+    class_info->head.parent = NULL;
+    func_80420FBC(class_info, 0, 1);
+}
+
+// (name based on KAR address)
+void func_804211B4(const char* name, HSD_ClassInfo* class_info)
```

Comment:
Left these comments on functions I don't know the names of, which exist in KAR but are stripped in melee because their only uses are auto-inlined. Maybe Psi knows what to call them.

Reply:
Maybe rename them to `kar_804211B4` for now so people don't think it's a Melee address?

Reply:
I'll just make a new issue so this can be merged.